### DISCUSSION TITLE
Create a checkbox from a variable-length list of Strings.

### DISCUSSION THREAD

**Discussion by jtreagan** [fltk-rs/fltk-rs#1584] [Q&A]
What I'm trying to do is create a checkbox widget that will display the names of the files in a directory (one button each) and then allow the user to choose which of those files he/she wants to use, returning those chosen file names in a vector of strings.  I started with the example given in the FLTK Book and that works fine for me.  The issue I can't seem to get past is that the number of filenames being passed to the widget is variable.  It might be a list of 4 strings or maybe 10.  That means that I won't know ahead of time how many buttons I need to create to be displayed in the widget.  The AI associated with my IDE had the bright idea of using a loop to run through the list of file name strings creating a button for each string, then storing those buttons in a hashmap that looked like this:  (String, button) or (file_name, button) .  I like that idea, but then implementing the display of those buttons in the widget stumped it.  Any suggestions?

**Comment by MoAlyousef** [comment]
Hello

I'm not sure I completely understand what you're trying to do, but I wouldn't store widgets in a hashmap. It just complicates things.

You can always generate widgets using a loop as the AI suggested. You can also iterate the children of a widget (casting them if needed), to get back info. Regenerate using a loop etc.
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    // in your app, a dynamic vec which gets files from the filesystem
    let files = vec!["file1.jpg", "file2.png", "file3.svg"];

    let app = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    let mut row = group::Flex::default_fill().row();
    let mut scroll = group::Scroll::default();
    row.fixed(&scroll, 100);
    let mut pack = group::Pack::default().with_size(100, 300);
    for file in files {
 	let check = button::CheckButton::default().with_label(file).with_size(0, 30);
    }
    pack.end();
    scroll.end();
    let mut btn = button::Button::default().with_label("@>");
    row.fixed(&btn, 30);
    let mut output = output::MultilineOutput::default();
    row.end();
    win.end();
    win.show();

    btn.set_callback(move |_b| {
	output.set_value("");
	let mut string = String::new();
	for i in 0..pack.children() {
		let check: button::CheckButton = button::CheckButton::from_dyn_widget(&pack.child(i).unwrap()).unwrap();
		if check.value() { // you can also use is_checked()
			string.push_str(&check.label());
			string.push('\n');
		}
	}
	output.set_value(&string);
    });
    app.run().unwrap();
}
```

**Reply by jtreagan** [reply]
Thank you, @MoAlyousef.  That is really beautiful and much nicer than what I envisioned.  I took the time to enter your code manually (rather than copy/paste) so that I could, maybe, learn what the code is doing.  It helped, but I'm going to have to dig into it more.  You can probably expect some more followup questions.

Here's my first followup question:  What is the best way to extract the gathered data from the `btn.set_callback(move |_b| {` closure?  I've run into this before, when working with menus, and would really like to find a good, workable solution.  In this case, the data is stored in the `string` variable within the closure, but `string` goes out of scope when the closure ends.  I need to access that data for use elsewhere in my program.  What's the best way for me to go about extracting it?

Thanks for your help.  I really appreciate it.

**Reply by MoAlyousef** [reply]
You're welcome.
You'll need a reference type to be able to store data from inside a callback.
Rust's lifetime rules requires that the reference outlive the callback. For that you can use a smart pointer with interior mutability. Something like `Rc<RefCell<SomeType>>` or `Arc<Mutex<SomeType>>` for multithreaded code.

You can use globals (statics) wrapped in mutices as well.

An example:
```rust
use fltk::{app, button::Button, prelude::*, window::Window};
use std::cell::RefCell;
use std::rc::Rc;

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(200, 200);
    let mut btn = Button::default_fill().with_label("Click");
    wind.end();
    wind.show();

    let count: i32 = 0;
    let count_rc = Rc::from(RefCell::from(count));

    btn.set_callback({
        let count_rc = count_rc.clone();
        move |_| *count_rc.borrow_mut() += 1
    });

    app::add_timeout(3.0, move || {
        println!(
            "At 3 seconds, the button was clicked {} times",
            *count_rc.borrow()
        )
    });

    app.run().unwrap();
}
```

An example of a global:
```rust
use fltk::{app, button::Button, prelude::*, window::Window};
use std::sync::Mutex;

static COUNT: Mutex<i32> = Mutex::new(0);

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(200, 200);
    let mut btn = Button::default_fill().with_label("Click");
    wind.end();
    wind.show();

    btn.set_callback(
        move |_| *COUNT.lock().unwrap() += 1
    );

    app::add_timeout(3.0, move || {
        println!(
            "At 3 seconds, the button was clicked {} times",
            *COUNT.lock().unwrap()
        )
    });

    app.run().unwrap();
}
```

fltk-rs also provides a singleton GlobalState type which can be modified inside a callback and is still accessible outside it:
```rust
use fltk::{app, button::Button, prelude::*, window::Window};

struct MyState {
    count: i32,
}

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(200, 200);
    let mut btn = Button::default_fill().with_label("Click");
    wind.end();
    wind.show();

    let initial_state = MyState { count: 0 };
    let _state = app::GlobalState::new(initial_state);

    btn.set_callback(
        move |_| app::GlobalState::<MyState>::get().with(|c| c.count += 1)
    );

    app::add_timeout(3.0, move || {
        println!(
            "At 3 seconds, the button was clicked {} times",
            app::GlobalState::<MyState>::get().with(|c| c.count)
        )
    });

    app.run().unwrap();
}
```
